consult.el

```diff
@@ -3010,48 +3015,72 @@ INITIAL is the initial input."
 	     (unless (or (bolp) (eobp))
 	       (forward-line 0))
 	     (point))))
-        (consult--each-line beg end
-          (push (buffer-substring-no-properties beg end) lines)
-          (push (make-overlay beg (1+ end)) overlays))))
-    (unless (use-region-p)
-      (goto-char (point-min)))
-    (lambda (input restore)
-      ;; New input provided -> Update
-      (when (and input (not (equal input last-input)))
-        (if (string-match-p "\\`!? ?\\'" input)
-            ;; Special case the empty input for performance.
-            (progn
-              (dolist (ov overlays)
-                (overlay-put ov 'invisible nil))
-              (setq last-input input))
-          (let* ((not (string-prefix-p "! " input))
-                 (stripped (string-remove-prefix "! " input))
-                 ;; Heavy computation is interruptible if *not* committing!
-                 (ht (if restore
-                         (consult--string-hash (funcall filter stripped lines))
-                       (while-no-input
-                         (consult--string-hash (funcall filter stripped lines))))))
-            (when (hash-table-p ht)
-              (let ((ov overlays) (li lines))
-                (while ov
-                  (overlay-put (car ov) 'invisible (eq not (gethash (car li) ht)))
-                  (setq li (cdr li) ov (cdr ov))))
-              (setq last-input input)))))
-      (when restore
-        (cond
-         ((not input)
-          (goto-char point-orig))
-         ((equal input "")
-          (consult-focus-lines 'show))
-         (t
-          ;; Sucessfully terminated -> Remember invisible overlays
-          (dolist (ov overlays)
-            (if (overlay-get ov 'invisible)
-                (push ov consult--focus-lines-overlays)
-              (delete-overlay ov)))
-          (setq overlays nil)))
-        ;; Destroy remaining overlays
-        (mapc #'delete-overlay overlays)))))
+        (let ((i 0) line)
+	  (consult--each-line beg end
+	    (setq line
+		  (if (eq beg end) (char-to-string ?\n) ; "\n" reuses string! 
+		    (buffer-substring-no-properties beg end)))
+	    (add-text-properties 0 1 `(line (,(cl-incf i) ,beg ,end)) line)
+	    (push line lines)))))
+    (cl-labels ((add-inv-ov (beg end)
+			    (push (consult--overlay beg end 'invisible t)
+				  overlays)))
```

Comment:
This function should be inlined. I rarely use cl-labels and in this case it is not justified, since the function is tiny.

Reply:
It's used 4 times in the potentially hard to parse and understand core algorithm, so I thought it would be easier for future readers.

Reply:
It is still only a single line and I don't consider this a "logical functional unit", which deserves to be factored out. Also it potentially makes the code slower. In other parts of Consult I use a functional programming style heavily, e.g., check out the async pipeline, but from my experience it does not feel justified here.

Reply:
I understand.  Is a `defsubst` also out of the question?